src/main/java/mekanism/common/integration/computer/ComputerMethodFactory.java

```diff
@@ -0,0 +1,107 @@
+ * These will then be tested and, if not restricted, "bound" to the holder in {@link #bindTo(Object, BoundMethodHolder)}
+ * Methods with the same name must have different parameter counts.
+ *
+ * @param <T> the "subject" that this Factory's methods operate on.
+ */
+@ParametersAreNotNullByDefault
+public class ComputerMethodFactory<T>{
+    protected static String[] NO_STRINGS = new String[0];
+    protected static Class<?>[] NO_CLASSES = new Class[0];
+
+    protected static MethodHandles.Lookup lookup = MethodHandles.lookup();
+
+    @SuppressWarnings("unused")//may be used by annotation processor
+    protected static MethodHandle getMethodHandle(Class<?> containingClass, String methodName, Class<?>... params) {
+        try {
+            Method method = containingClass.getDeclaredMethod(methodName, params);
+            method.setAccessible(true);
+            return lookup.unreflect(method);
+        } catch (ReflectiveOperationException roe) {
+            throw new RuntimeException("Couldn't get method handle for "+methodName, roe);
+        }
+    }
+
+    @SuppressWarnings("unused")//may be used by annotation processor
+    protected static MethodHandle getGetterHandle(Class<?> containingClass, String fieldName) {
+        try {
+            Field field = containingClass.getDeclaredField(fieldName);
+            field.setAccessible(true);
+            return lookup.unreflectGetter(field);
+        } catch (ReflectiveOperationException roe) {
+            throw new RuntimeException("Couldn't get getter MethodHandle for "+fieldName, roe);
+        }
+    }
+
+    private final List<MethodData<T>> methods = new ArrayList<>();
+    /**
+     * Method + arg count pairs to make sure methods are unique
+     */
+    private final Set<ObjectIntPair<String>> methodsKnown = new HashSet<>();
+
+    protected void register(String name, MethodRestriction restriction, String[] requiredMods, boolean threadSafe, Class<?> returnType, ComputerFunctionCaller<T> handler, String[] argumentNames, Class<?>[] argClasses) {
+        if (!methodsKnown.add(new ObjectIntImmutablePair<>(name, argumentNames.length))) {
+            throw new RuntimeException("Duplicate method name "+name+"_"+argumentNames.length);
+        }
+        this.methods.add(new MethodData<>(name, restriction, requiredMods, threadSafe, argumentNames, argClasses, returnType, handler));
+    }
+
+    protected void register(String name, MethodRestriction restriction, String[] requiredMods, boolean threadSafe, Class<?> returnType, ComputerFunctionCaller<T> handler) {
+        register(name, restriction, requiredMods, threadSafe, returnType, handler, NO_STRINGS, NO_CLASSES);
+    }
+
+    void bindTo(@Nullable T subject, BoundMethodHolder holder) {
+        WeakReference<T> weakSubject = subject != null ? new WeakReference<>(subject) : null;
+        for (MethodData<T> methodData : this.methods) {
+            if (methodData.restriction.test(subject) && modsLoaded(methodData.requiredMods)) {
```

Comment:
We may want to just move this check of restrictions matching AND required mods being loaded into a helper method in `MethodData`

Reply:
I think it's fine here, we're not going to do the check anywhere else